### DISCUSSION TITLE
dynamic learning rate

### DISCUSSION THREAD

**Discussion by ANAS1997AA** [ultralytics/ultralytics#8874] [Docs]
hello everyone
I wanted to train a new model with dynamic learning rate in yolov8n and I did not find any solution for that. any help please ?

**Comment by pderrenger** [comment]
Hello! 👋

Great question! In YOLOv8, you can adjust the learning rate dynamically during training using the `--lr0` (initial learning rate), `--lrf` (final learning rate factor), and scheduler settings in your training command or YAML configuration file. The learning rate scheduler adjusts the learning rate based on these settings and the number of epochs.

For example, to start with an initial learning rate of 0.01 and reduce it to 10% of its initial value by the end of training, you can use:

```bash
yolo detect train data=coco128.yaml model=yolov8n.yaml epochs=100 lr0=0.01 lrf=0.1
```

This sets up a dynamic learning rate that starts at 0.01 and decreases to 0.001 (0.01 * 0.1) by the end of training, following the scheduler's policy.

Hope this helps! Let us know if you have any more questions. 😊

**Reply by ANAS1997AA** [reply]
this is my command !yolo task=detect mode=train model=yolov8n.pt data={dataset.location}/anas.yml epochs=50 imgsz=800 plots=True batch=16 iou=0.45 lr0=0.001 lrf=0.01 
and the output is 
![Capture d'écran 2024-03-12 124159](https://github.com/ultralytics/ultralytics/assets/162803130/566bedee-c1ea-459a-bd0b-34a413f1f677)

the optimizer and the learning rate are set to auto. what can I do

**Reply by pderrenger** [reply]
Hello! 👋

It looks like you're doing great with your command setup! If the optimizer and learning rate are set to auto and you wish to specify them manually, you can directly set the optimizer in your `anas.yml` under the `optimizer` field. For the learning rate, you've already specified `lr0` and `lrf`, which should override the auto settings.

If you're still seeing 'auto' despite this, ensure your `anas.yml` doesn't reset these values. If the issue persists, it might be helpful to share a snippet of your `anas.yml` to better diagnose the problem.

Here's a quick example of specifying an optimizer in your YAML:

```yaml
# anas.yml
optimizer: SGD  # or 'Adam', 'AdamW', etc.
```

Keep up the good work, and let us know how it goes! 😊

**Reply by ANAS1997AA** [reply]
I want to do the train with an exponential decay in the learning rate. how cazn i do it ?

**Reply by pderrenger** [reply]
Hello! 😊

To implement an exponential decay in the learning rate during training with YOLOv8, you can adjust the `lrf` (learning rate factor) in your training command. The `lrf` controls how the learning rate decreases over the course of training. For an exponential decay, set `lrf` to a value less than 1. Here's an example:

```bash
yolo detect train model=yolov8n.pt data=your_dataset.yaml epochs=100 lr0=0.01 lrf=0.01
```

This setup starts with an initial learning rate (`lr0`) of 0.01 and reduces it exponentially to 1% of its initial value by the end of training. Adjust `lrf` as needed for your specific requirements!

Hope this helps! Let us know if you have any more questions.

**Reply by ANAS1997AA** [reply]
i didn't understand the example well

**Reply by pderrenger** [reply]
Sure, I'd be happy to clarify! 😊 If you're looking to implement an exponential decay for the learning rate in YOLOv8, you can do so by setting the `lr0` (initial learning rate) and `lrf` (learning rate factor) in your training command. Here's a simplified example:

```bash
yolo detect train model=yolov8n.pt data=your_data.yaml epochs=50 lr0=0.01 lrf=0.01
```

This means your learning rate starts at 0.01 and decreases exponentially to 0.01 * 0.01 by the end of 50 epochs. Adjust `lr0` and `lrf` to fit your needs. Hope this helps!
